pkg/gateway/multiparts/tracker.go

```diff
@@ -38,82 +72,70 @@ var (
+}
+
+func protoFromMultipart(m *MultipartUpload) *MultipartUploadData {
+	return &MultipartUploadData{
+		UploadId:        m.UploadID,
+		Path:            m.Path,
+		CreationDate:    timestamppb.New(m.CreationDate),
+		PhysicalAddress: m.PhysicalAddress,
+		Metadata:        m.Metadata,
+		ContentType:     m.ContentType,
 	}
 }
 
 func (m *tracker) Create(ctx context.Context, multipart MultipartUpload) error {
 	if multipart.UploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		_, err := tx.Exec(`INSERT INTO gateway_multiparts (upload_id,path,creation_date,physical_address, metadata, content_type)
-			VALUES ($1, $2, $3, $4, $5, $6)`,
-			multipart.UploadID, multipart.Path, multipart.CreationDate, multipart.PhysicalAddress, multipart.Metadata, multipart.ContentType)
-		return nil, err
-	})
+	_, err := m.Get(ctx, multipart.UploadID)
+	if err == nil {
+		return ErrAlreadyExists
+	}
+	path := multipartPrefix + kv.PathDelimiter + multipart.UploadID
+	err = m.store.SetMsg(ctx, path, protoFromMultipart(&multipart))
 	return err
 }
 
 func (m *tracker) Get(ctx context.Context, uploadID string) (*MultipartUpload, error) {
 	if uploadID == "" {
 		return nil, ErrInvalidUploadID
 	}
-	res, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		var m MultipartUpload
-		if err := tx.Get(&m, `
-			SELECT upload_id, path, creation_date, physical_address, metadata, content_type 
-			FROM gateway_multiparts
-			WHERE upload_id = $1`,
-			uploadID); err != nil {
-			return nil, err
-		}
-		return &m, nil
-	})
+	data := &MultipartUploadData{}
+	path := multipartPrefix + kv.PathDelimiter + uploadID
+	err := m.store.GetMsg(ctx, path, data)
 	if err != nil {
 		return nil, err
 	}
-	return res.(*MultipartUpload), nil
+	return multipartFromProto(data), nil
 }
 
 func (m *tracker) Delete(ctx context.Context, uploadID string) error {
 	if uploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		res, err := tx.Exec(`DELETE FROM gateway_multiparts WHERE upload_id = $1`, uploadID)
-		if err != nil {
-			return nil, err
-		}
-		affected := res.RowsAffected()
-		if affected != 1 {
-			return nil, ErrMultipartUploadNotFound
+	if _, err := m.Get(ctx, uploadID); err != nil {
+		if errors.Is(err, kv.ErrNotFound) {
+			return ErrMultipartUploadNotFound
 		}
-		return nil, nil
-	})
-	return err
-}
-
-func (m Metadata) Set(k, v string) {
-	m[strings.ToLower(k)] = v
-}
-
-func (m Metadata) Get(k string) string {
-	return m[strings.ToLower(k)]
-}
-func (m Metadata) Value() (driver.Value, error) {
-	return json.Marshal(m)
-}
-
-func (m *Metadata) Scan(src interface{}) error {
-	if src == nil {
-		return nil
+		return fmt.Errorf("failed on Get. Key (%s): %w", uploadID, err)
 	}
-	data, ok := src.([]byte)
-	if !ok {
-		return ErrInvalidMetadataSrcFormat
-	}
-	return json.Unmarshal(data, m)
+	path := multipartPrefix + kv.PathDelimiter + uploadID
```

Comment:
Not part of this PR - I think that one of the utilities of the KV package should be concatenating paths. We messed up with it so many times with the block adapters..

Reply:
Done